### PR TITLE
Make exhaustiveness usable outside of rustc

### PR DESCRIPTION
**Pull request by Nadrieril**
With this PR, `rustc_pattern_analysis` compiles on stable (with the `stable` feature)! `rust-analyzer` will be able to use it to provide match-related diagnostics and refactors.

Two questions:
- Should I name the feature `nightly` instead of `rustc` for consistency with other crates? `rustc` makes more sense imo.
- `typed-arena` is an optional dependency but tidy made me add it to the allow-list anyway. Can I avoid that somehow?

r? @compiler-errors

### CODE REVIEW

**Comment by Nadrieril** [comment]
`typed-arena` is MIT licensed btw

**Comment by Nadrieril** [comment]
Bracing myself for perf sadness

@bors try @rust-timer queue

File: compiler/rustc_pattern_analysis/src/constructor.rs [hunk 1]
```diff
@@ -866,13 +874,27 @@ impl ConstructorSet {
...
+                }
+            }
+            ConstructorSet::Ref => {
+                if !seen.is_empty() {
+                    present.push(Ref);
+                } else {
+                    missing.push(Ref);
+                }
+            }
+            ConstructorSet::Union => {
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/constructor.rs:901]: Can unions be uninhabited? 🤔

File: compiler/rustc_pattern_analysis/src/pat.rs [hunk 2]
```diff
@@ -90,31 +89,25 @@ impl<'p, 'tcx> DeconstructedPat<'p, 'tcx> {
...
-                            .cx
-                            .pattern_arena
-                            .alloc(DeconstructedPat::wildcard(inner_ty, DUMMY_SP));
-                        let extra_wildcards = other_slice.arity() - self_slice.arity();
-                        let extra_wildcards = (0..extra_wildcards).map(|_| wildcard);
-                        prefix.iter().chain(extra_wildcards).chain(suffix).collect()
-                    }
+            (
+                &Slice(self_slice @ Slice { kind: SliceKind::VarLen(prefix, suffix), .. }),
+                &Slice(other_slice),
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/pat.rs:98]: Is there sorting or something that ensures the smaller slice is `self_slice`?

File: compiler/rustc_pattern_analysis/src/lib.rs [hunk 3]
```diff
@@ -79,6 +79,13 @@ pub struct MatchArm<'p, Cx: MatchCx> {
 
 impl<'p, Cx: MatchCx> Copy for MatchArm<'p, Cx> {}
 
+#[cfg(not(feature = "rustc"))]
+pub trait Captures<'a> {}
+#[cfg(not(feature = "rustc"))]
+impl<'a, T: ?Sized> Captures<'a> for T {}
+#[cfg(feature = "rustc")]
+pub use rustc_data_structures::captures::Captures;
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/lib.rs:87]: We should just unconditionally duplicate the `Captures` trait here. No reason to import it from `rustc_data_structures`, it's small enough that code repetition is not an issue.

File: compiler/rustc_pattern_analysis/src/constructor.rs [hunk 4]
```diff
@@ -266,6 +266,7 @@ pub struct IntRange {
 
 impl IntRange {
     /// Best effort; will not know that e.g. `255u8..` is a singleton.
+    #[cfg_attr(not(feature = "rustc"), allow(dead_code))]
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/constructor.rs:269]: Why is this gated?

File: compiler/rustc_pattern_analysis/src/rustc.rs [hunk 5]
```diff
@@ -24,21 +24,19 @@ use crate::MatchCx;
...
 pub type Usefulness = crate::usefulness::Usefulness<Span>;
-pub type UsefulnessReport<'p, 'tcx> =
-    crate::usefulness::UsefulnessReport<'p, MatchCheckCtxt<'p, 'tcx>>;
-pub type WitnessPat<'p, 'tcx> = crate::pat::WitnessPat<MatchCheckCtxt<'p, 'tcx>>;
+pub type UsefulnessReport<'p, 'tcx> = crate::usefulness::UsefulnessReport<'p, RustcCtxt<'p, 'tcx>>;
+pub type WitnessPat<'p, 'tcx> = crate::pat::WitnessPat<RustcCtxt<'p, 'tcx>>;
 
 #[derive(Clone)]
-pub struct MatchCheckCtxt<'p, 'tcx> {
+pub struct RustcCtxt<'p, 'tcx> {
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/rustc.rs:40]: `RustcMatchCheckCtxt`? Kinda an arm-full, but seems a bit more relevant name?

File: compiler/rustc_pattern_analysis/src/lib.rs [hunk 6]
```diff
@@ -39,17 +39,19 @@ pub fn analyze_match<'p, 'tcx>(
...
 
     // Lint on ranges that overlap on their endpoints, which is likely a mistake.
-    lint_overlapping_range_endpoints(cx, &pat_column);
+    lint_overlapping_range_endpoints(cx, &pat_column, wildcard_arena);
 
     // Run the non_exhaustive_omitted_patterns lint. Only run on refutable patterns to avoid hitting
     // `if let`s. Only run if the match is exhaustive otherwise the error is redundant.
     if cx.refutable && report.non_exhaustiveness_witnesses.is_empty() {
-        lint_nonexhaustive_missing_variants(cx, arms, &pat_column, scrut_ty)
+        lint_nonexhaustive_missing_variants(cx, arms, &pat_column, scrut_ty, wildcard_arena)
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/lib.rs:110]: Why do we pass in both the context and the arena?

File: compiler/rustc_pattern_analysis/src/lints.rs [hunk 7]
```diff
@@ -27,11 +28,11 @@ use crate::MatchArm;
 ///
 /// This is not used in the main algorithm; only in lints.
 #[derive(Debug)]
-pub(crate) struct PatternColumn<'p, 'tcx> {
-    patterns: Vec<&'p DeconstructedPat<'p, 'tcx>>,
+pub(crate) struct PatternColumn<'a, 'p, 'tcx> {
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/lints.rs:31]: Is the `'p` lifetime now invariant? Why did you need to introduce a new lifetime here?

File: compiler/rustc_pattern_analysis/src/lints.rs [hunk 8]
```diff
@@ -80,7 +81,14 @@ impl<'p, 'tcx> PatternColumn<'p, 'tcx> {
     /// This returns one column per field of the constructor. They usually all have the same length
     /// (the number of patterns in `self` that matched `ctor`), except that we expand or-patterns
     /// which may change the lengths.
-    fn specialize(&self, pcx: &PatCtxt<'_, 'p, 'tcx>, ctor: &Constructor<'tcx>) -> Vec<Self> {
+    fn specialize<'b>(
+        &self,
+        pcx: &'b PatCtxt<'_, 'p, 'tcx>,
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/lints.rs:86]: `&'b PatCtxt<'a, 'p, 'tcx>`?

File: compiler/rustc_pattern_analysis/src/pat.rs (hunk 2 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/pat.rs:98]: Yep, because we only run `specialize` when `other_ctor` is covered by `self.ctor`, which for slices means that `self_slice` is smaller

File: compiler/rustc_pattern_analysis/src/constructor.rs (hunk 4 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/constructor.rs:269]: I wanted it to compile without warnings with either feature, I could also just make it `pub`

File: compiler/rustc_pattern_analysis/src/constructor.rs (hunk 1 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/constructor.rs:901]: Not according to `inhabited_predicate`. There's a comment "For now, unions are always considered inhabited" there which I'm inclined to trust

Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/constructor.rs:901]: Validity for unions is pretty undecided territory

File: compiler/rustc_pattern_analysis/src/lib.rs (hunk 6 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/lib.rs:110]: The alternative would be to only pass the context and add the arena stuff to the `MatchCx` trait, but the only way I managed that required a lifetime on the trait which infected everything. I expect I'll eventually use a `MatchCheckCtxt<Cx>` that carries the arena, the context, and other context-agnostic things

File: compiler/rustc_pattern_analysis/src/lints.rs (hunk 7 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/lints.rs:31]: Yeah, it's invariant because `SmallVec<[T; N]>` is invariant in `T`

Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/lints.rs:31]: Honestly I'm not exactly sure, I was happy to find something that worked, this took a while to untangle

Comment (compiler-errors) [reply compiler/rustc_pattern_analysis/src/lints.rs:31]: Darn

File: compiler/rustc_pattern_analysis/src/constructor.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_pattern_analysis/src/constructor.rs:269]: Well if it's unused with `not(feature = "rustc")`, then it's actually unused, and shouldn't we just cfg it out? If you expect it to be public for things like r-a, then make it pub.

File: compiler/rustc_pattern_analysis/src/lints.rs (hunk 8 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/lints.rs:86]: Oh yeah

File: compiler/rustc_pattern_analysis/src/lib.rs (hunk 6 above)
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/lib.rs]: Well I did just that, and cleaned up `MatchCx` and `PatCtxt` while I was at it

**Comment by compiler-errors** [comment]
I will review this on Monday probably

File: compiler/rustc_pattern_analysis/src/lib.rs [hunk 9]
```diff
@@ -1,54 +1,133 @@
...
+#[cfg(feature = "stable")]
+pub(crate) use typed_arena::Arena as TypedArena;
+
+pub trait Captures<'a> {}
+impl<'a, T: ?Sized> Captures<'a> for T {}
+
+/// Context that provides type information about constructors.
+///
+/// Most of the crate is parameterized on a type that implements this trait.
+pub trait TypeCx: Sized + Clone + fmt::Debug {
```
Comment (compiler-errors) [inline compiler/rustc_pattern_analysis/src/lib.rs:52]: You shouldn't need a `Clone` supertrait here.

**Review by compiler-errors — APPROVED** [review]
Please address these in a follow-up commit

File: compiler/rustc_index/src/bit_set.rs [hunk 10]
```diff
@@ -491,10 +493,21 @@ impl<T: Idx> ChunkedBitSet<T> {
...
             Zeros(chunk_domain_size) => {
                 if chunk_domain_size > 1 {
-                    // We take some effort to avoid copying the words.
-                    let words = Rc::<[Word; CHUNK_WORDS]>::new_zeroed();
-                    // SAFETY: `words` can safely be all zeroes.
-                    let mut words = unsafe { words.assume_init() };
+                    #[cfg(feature = "nightly")]
+                    let mut words = {
+                        // We take some effort to avoid copying the words.
+                        let words = Rc::<[Word; CHUNK_WORDS]>::new_zeroed();
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_index/src/bit_set.rs:499]: @Nadrieril should add this to a follow-up PR if it matters.

File: compiler/rustc_pattern_analysis/src/lib.rs [hunk 11]
```diff
@@ -1,54 +1,133 @@
...
+
+    /// Best-effort `Debug` implementation.
+    fn debug_pat(f: &mut fmt::Formatter<'_>, pat: &DeconstructedPat<'_, Self>) -> fmt::Result;
+
+    /// Raise a bug.
+    fn bug(&self, fmt: fmt::Arguments<'_>) -> !;
+}
+
+/// Context that provides information global to a match.
+#[derive(Clone)]
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/lib.rs:88]: Oh nice I didn't know we used `derivative`

**Comment by Kobzol** [comment]
Small regression to a single stress test benchmark. Worth the librarification effort, I don't think that we need to dive further into this.

@rustbot label: +perf-regression-triaged
